pkg/timelock/timelock.go

```diff
@@ -240,54 +258,73 @@ func (tw *Worker) Listen(ctx context.Context) error {
 				// Check if the error is not nil, because sub.Unsubscribe will
 				// signal the channel sub.Err() to close it, leading to false nil errors.
 				if err != nil {
-					tw.logger.Info().Msgf("subscription: %s", err.Error())
+					tw.logger.Warn().Msgf("subscription error: %s", err.Error())
 					SetReadyStatus(HealthStatusError)
-					loop = false
+					sub.Unsubscribe()
+
+					success := false
+					for try := range maxSubRetries {
+						tw.logger.Warn().Msgf("trying to re-create subscription: %v/%v retry.", try+1, maxSubRetries)
+						sub, err = tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
+						if err == nil {
+							tw.logger.Info().Msg("subscription successfully recreated.")
+							SetReadyStatus(HealthStatusOK)
+							success = true
+							break
+						}
+
+						time.Sleep(time.Second * time.Duration(try))
+					}
+
+					if !success {
+						tw.logger.Error().Msg("failed to recreate subscription after retries: shutting down timelock-worker.")
+						break MainLoop
+					}
 				}
 
 			case signal := <-stopCh:
 				tw.logger.Info().Msgf("received OS signal %s", signal)
 				SetReadyStatus(HealthStatusError)
-				loop = false
+				break MainLoop
 			}
 		}
 		wg.Done()
```

Comment:
`defer wg.Done()` before you start the main loop and you don't need to label the loop

Reply:
Applying it.